### DISCUSSION TITLE
Array support

### DISCUSSION THREAD

**Discussion by ubIQio** [microhomie/microhomie#51] [General]
New to microhomie, and am trying to implement a node that generates periodic arrays (FFT of audio data)

So I'd like to be able construct a property that could be described as an array or list of integers for display by an MQTT subscriber.

I see that arrays have been removed from node properties in recent version as there was 'no use case'.

What would be the recommended way to support this use case?

**Comment by rroemhild** [comment]
Arrays are no longer support by the Homie convention since version 4, so we removed the support too.

I have a side project where I implemented array support for. I'm just not sure if this is a good idea/solution or if anybody wants it, but it works for me :smile: But I'm glad you asked, this gives me motivation to think about to integrate it. For now I will create a branch in the next days (give me some time, high workload), where you can test my array solution if it fits your problem and we can discuss an array node.

The correct Homie way would be to write an array extension specification to describe the array function. But we can go this way if we have a working solution.

**Reply by ubIQio** [reply]
Thanks, thats really helpful.
For now I just publish as a string, but I do think it would be helpful to subscribers to know that the type is array so that the data display can be formatted appropriately

**Comment by rroemhild** [comment]
I've merged and published my array node/property code to the [feature/arraynode](https://github.com/microhomie/microhomie/tree/feature/arraynode) branch. There is also an example in `examples/array` in this branch. My idea was to use a sub-topic for each array index.

It also subscribes to the `<propderty_id>/set topic` to i.e. turn all leds on or off.

The output from the example array device is as follows:

```
MQTT SUBSCRIBE: homie/208b0300/led/+/power
MQTT SUBSCRIBE: homie/208b0300/led/power/set
MQTT SUBSCRIBE: homie/208b0300/led/+/power/set

MQTT PUBLISH: homie/208b0300/$name --> b'Array node test'
MQTT PUBLISH: homie/208b0300/$nodes --> b'led'
MQTT PUBLISH: homie/208b0300/led/$name --> b'LED'
MQTT PUBLISH: homie/208b0300/led/$array --> b'16'
MQTT PUBLISH: homie/208b0300/led/$type --> b'LED'
MQTT PUBLISH: homie/208b0300/led/$properties --> b'power'
MQTT PUBLISH: homie/208b0300/led/$name --> b'Power'
MQTT PUBLISH: homie/208b0300/led/$datatype --> b'boolean'
MQTT PUBLISH: homie/208b0300/led/$settable --> b'true'

MQTT UNSUBSCRIBE: homie/208b0300/led/+/power

MQTT PUBLISH: homie/208b0300/led/0/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/1/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/2/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/3/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/4/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/5/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/6/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/7/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/8/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/9/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/10/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/11/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/12/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/13/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/14/power --> b'true'
MQTT PUBLISH: homie/208b0300/led/15/power --> b'true'
```

Let me know if this address your problem.

Best regards
Rafael

**Reply by ubIQio** [reply]
Thanks so much for this, I think it will be useful, though in the end for speed/efficiency reasons I ended up send the FFT data as binary (its 2048 bytes every second)
I created a new data type of BINARY as a hint to the display app , but of course the content packing is application dependent , so not sure how to communicate that
